### ISSUE TITLE
UmiAwareMarkDuplicatesWithMateCigar assigns different duplicate flags for reads from the same fragment

### ISSUE THREAD

**Issue by yangruialex** [broadinstitute/picard#1449]
### Affected tool(s)
picard UmiAwareMarkDuplicatesWithMateCigar

### Affected version(s)
- [ ] Latest public release version [2.21.6]

### Description 
When multiple fragments are present for a single UMI family, picard UmiAwareMarkDuplicatesWithMateCigar does not always set duplicate flag per read pair, which leads to different duplicate flag for read1 and read2 from the same fragment. 

This issue is specific to the UMI aware tool. 

#### Steps to reproduce
This issue can be found by iterating through the read pairs and count the occurrence of `r1.is_duplicate != r2.is_duplicate` in `pysam`. 

An example bam file is provided [here](https://drive.google.com/open?id=1HJbZgsSDOQjHp0n2-9fdNe3ZuoP_Ceo1):

code I used to mark duplicates: 

#### Expected behavior
_Tell us what should happen_

#### Actual behavior
_Tell us what happens instead_

**Comment by yfarjoun** [comment]
@fleharty are you available to take a look at this?

**Comment by fleharty** [comment]
@yangruialex Thanks for the sample bam, could you post your pysam code that you used to identify this to help me identify the source of this apparent bug?

**Comment by fleharty** [comment]
@yangruialex Sorry, I don't need the pysam code, I have a way of looking that might actually be better in that it will let me identify read pairs that are differentially duplicate marked.

**Comment by yangruialex** [comment]
@fleharty Thank you for looking into this. I'll post my code anyways just to make sure it is not something on my end:

```
import itertools
import pysam


def iter_read_pair(reads):
    reads.sort(key=lambda x: x.query_name)
    for _, read_pair in itertools.groupby(reads, key=lambda x: (x.query_name)):
        yield tuple(read_pair)


def count_disagreement(bam_path):
    reads = list(pysam.AlignmentFile(bam_path).fetch('chr21'))

    total_read_pair = 0
    disagreement_read_pair = 0

    for read_pair in iter_read_pair(reads):
        if len(read_pair) == 2:
            total_read_pair += 1
            r1, r2 = read_pair
            assert type(r1.is_duplicate) == bool
            assert type(r2.is_duplicate) == bool
            if r1.is_duplicate != r2.is_duplicate:
                disagreement_read_pair += 1
    print("total read pair:", total_read_pair)
    print("read pairs with different dup flag:", disagreement_read_pair)
    print("fraction:", disagreement_read_pair/ total_read_pair)
```

**Comment by fleharty** [comment]
@yangruialex 
Is it possible you gave me the wrong example bam file?  I get the same number of read pairs with your tool, but 0 different dup flag....

13:09:02:gsa5:tmp 1016 $ python dup.py
('total read pair:', 113560)
('read pairs with different dup flag:', 0)
('fraction:', 0)

**Comment by yfarjoun** [comment]
Perhaps you need to run Mark duplicates on it yourself, Mark?

**Comment by yangruialex** [comment]
Yes sorry I wasn't clear, the example bam is not duplicate marked.

**Comment by fleharty** [comment]
@yangruialex @yfarjoun 
Yup got it, thanks!  I now get similar though not identical results, probably due to differing versions of python or pysam.

('total read pair:', 113560)
('read pairs with different dup flag:', 19918)
('fraction:', 0)

**Comment by yangruialex** [comment]
Also some contexts the might help troubleshooting, the bam comes from methylation sequencing, and is generated by[ bwameth](https://github.com/brentp/bwa-meth)

**Comment by fleharty** [comment]
@yangruialex I notice there are TONS of G->A and C->T base changes, is that due to methylation?

**Comment by yangruialex** [comment]
> @yangruialex I notice there are TONS of G->A and C->T base changes, is that due to methylation?

Yes, unmethylated Cs are turned into Ts in the original template. Does that affect duplicate marking?

**Comment by fleharty** [comment]
@yangruialex It shouldn't affect duplicate marking.  At first when I looked at the data I thought I had the wrong reference because it was so messy, but it makes sense now given that it is bisulfite seq.

**Comment by hillst** [comment]
hi @fleharty , any update?

**Comment by yfarjoun** [comment]
it seems that this bug is related to multithreading race condition in https://github.com/samtools/htsjdk/issues/1516. Now that this has been identified, I think a fix will be forthcoming. 

Nice one!

**Comment by fleharty** [comment]
@yfarjoun I was hopeful that samtools/htsjdk#1516 would fix this.  I've applied that fix on this data, and unfortunately it doesn't seem to work.

I also tested the tool SimpleMarkDuplicates, which relies on the duplicate set iterator.  It seems that the issue here also exists in that tool which indicates the issue is probably in the duplicate set iterator, and hence htsjdk.

**Comment by ggydush-fn** [comment]
@fleharty hope all is well with you!! Curious if you ever figured this out!

**Comment by fleharty** [comment]
@ggydush-fn 
It's been quite some time since I last looked into this.  I believe the issue with this is how the DuplicateSetIterator in HTSJDK works.

If I remember correctly, there was something unusual about the reads that resulted in this issue.  I'll rerun the code to see what the issue was.
